src/main/java/org/terasology/ligthandshadow/barrier/MagicDomeSystem.java

```diff
@@ -28,119 +18,135 @@
+                if (temp.sub(new Vector3f(LASUtils.CENTER_RED_BASE_POSITION), new Vector3f()).length()
+                        < temp.sub(new Vector3f(LASUtils.CENTER_BLACK_BASE_POSITION), new Vector3f()).length()) {
+                    projectile.getComponent(LASTeamComponent.class).team = "red";
+                } else {
+                    projectile.getComponent(LASTeamComponent.class).team = "black";
+                }
+            }
+            return barrier.getComponent(MagicDome.class).team.equals(projectile.getComponent(LASTeamComponent.class).team);
+        }
+        return false;
+    }
 
+    public void destroyProjectile(Iterable<EntityRef> items) {
+        for (EntityRef item : items) {
+            if (item.getComponent(LocationComponent.class) != null) {
+                LocationComponent itemLocationComponent = item.getComponent(LocationComponent.class);
+                Vector3f pos = itemLocationComponent.getWorldPosition(new Vector3f());
+                for (EntityRef domeEntity : entityManager.getEntitiesWith(MagicDome.class, LocationComponent.class)) {
+                    LocationComponent domeLocationComponent = domeEntity.getComponent(LocationComponent.class);
+                    Vector3f domeCenter = domeLocationComponent.getWorldPosition(new Vector3f());
+                    pos.sub(domeCenter, df);
+                    float distance = df.length();
+                    if (distance > WORLD_RADIUS && isProjectileAllowedInsideBarrier(item, domeEntity)) {
+                        item.destroy();
+                    }
                 }
-                lastPos.set(pos);
             }
         }
     }
 
     @Override
     public void update(float delta) {
         updateDelta += delta;
-        if (updateDelta > 2.0f) {
-            for (EntityRef entity : entityManager.getEntitiesWith(MagicDome.class, AnimateRotationComponent.class)) {
-                AnimateRotationComponent rotationComponent = entity.getComponent(AnimateRotationComponent.class);
-
-                rotationComponent.yawSpeed = Math.clamp(-0.01f, 0.01f, rotationComponent.yawSpeed + random.nextFloat(-0.01f, 0.01f));
-                rotationComponent.pitchSpeed = Math.clamp(-0.01f, 0.01f, rotationComponent.pitchSpeed + random.nextFloat(-0.01f, 0.01f));
-                rotationComponent.rollSpeed = Math.clamp(-0.01f, 0.01f, rotationComponent.rollSpeed + random.nextFloat(-0.01f, 0.01f));
-
-                entity.saveComponent(rotationComponent);
-            }
+        if (updateDelta > 0.1f) {
+            // required to destroy the fireball sent by the staff
+            Iterable<EntityRef> fireballs = entityManager.getEntitiesWith(ExplodeComponent.class, LASTeamComponent.class);
+            destroyProjectile(fireballs);
+            // required to destroy the arrows and the spear
+            Iterable<EntityRef> arrows = entityManager.getEntitiesWith(ArrowComponent.class, LASTeamComponent.class);
+            destroyProjectile(arrows);
```

Comment:
It would be nice to have a `ProjectileComponent` marker component here, I guess...I don't think we have any real concept for hierarchies like that yet.